### ISSUE TITLE
Simplification pass does not take function call into account

### ISSUE THREAD

**Issue by k-ye** [taichi-dev/taichi#2442]
**Describe the bug**

Function inlining is a backend-independent process, so we should not really have to exclude opengl or cc backend. They failed because this test actually doesn't work. And it passed on LLVM backends only because assertion is only enabled when `debug=True`.

https://github.com/taichi-dev/taichi/blob/e1c1062bbbbfe9d6d4665a0d6fac750ddf0c3394/tests/python/test_function.py#L151-L186

Here's the IR before and after `simplication I`. Because the simplification didn't take into account the `call`, the load after the function call got incorrectly forwarded to constant `10`.

```
[I 06/20/21 01:00:19.898 4743931] [compile_to_offloads.cpp:operator()@22] [run_func_c6_0] Bit Loop Vectorized:
kernel {
  <i32> $0 = const [10]
  <*i32> $1 = global ptr [S2place<i32>], index [] activate=true
  $2 : global store [$1 <- $0]
  <i32> $3 = const [20]
  <*i32> $4 = global ptr [S4place<i32>], index [] activate=true
  $5 : global store [$4 <- $3]
  $6 = call "inc_13_0", args = {}
  <i32> $7 = const [1]
  <*i32> $8 = global ptr [S2place<i32>], index [] activate=true
  <i32> $9 = global load $8
  <i32> $10 = const [11]
  <i32> $11 = cmp_eq $9 $10
  <i32> $12 = bit_and $7 $11
  13 : assert $12, "(x[None] == 11)"
  <i32> $14 = const [1]
  <*i32> $15 = global ptr [S4place<i32>], index [] activate=true
  <i32> $16 = global load $15
  <i32> $17 = const [20]
  <i32> $18 = cmp_eq $16 $17
  <i32> $19 = bit_and $14 $18
  20 : assert $19, "(y[None] == 20)"
  $21 = call "inc_13_1", args = {}
  <i32> $22 = const [1]
  <*i32> $23 = global ptr [S2place<i32>], index [] activate=true
  <i32> $24 = global load $23
  <i32> $25 = const [11]
  <i32> $26 = cmp_eq $24 $25
  <i32> $27 = bit_and $22 $26
  28 : assert $27, "(x[None] == 11)"
  <i32> $29 = const [1]
  <*i32> $30 = global ptr [S4place<i32>], index [] activate=true
  <i32> $31 = global load $30
  <i32> $32 = const [21]
  <i32> $33 = cmp_eq $31 $32
  <i32> $34 = bit_and $29 $33
  35 : assert $34, "(y[None] == 21)"
}

[I 06/20/21 01:00:19.899 4743931] [compile_to_offloads.cpp:operator()@22] [run_func_c6_0] Simplified I:
kernel {
  <i32> $0 = const [10]
  <*i32> $1 = global ptr [S2place<i32>], index [] activate=true
  $2 : global store [$1 <- $0]
  <i32> $3 = const [20]
  <*i32> $4 = global ptr [S4place<i32>], index [] activate=true
  $5 : global store [$4 <- $3]
  $6 = call "inc_13_0", args = {}
  <i32> $7 = const [1]
  <i32> $8 = const [11]
  <i32> $9 = cmp_eq $0 $8  <--- !
  <i32> $10 = bit_and $9 $7
  11 : assert $10, "(x[None] == 11)"
  $12 = call "inc_13_1", args = {}
  13 : assert $10, "(x[None] == 11)"
  <i32> $14 = const [21]
  <i32> $15 = cmp_eq $3 $14
  <i32> $16 = bit_and $15 $7
  17 : assert $16, "(y[None] == 21)"
}
```

**Comment by xumingkuan** [comment]
Thanks for finding this bug! I'll take a look later.

**Comment by xumingkuan** [comment]
> Function inlining is a backend-independent process, so we should not really have to exclude opengl or cc backend. They failed because this test actually doesn't work.

Right, but we do not support assertions in these backends for now. I just opened #2447 for this.

> And it passed on LLVM backends only because assertion is only enabled when `debug=True`.

This should be fixed in #2448 :)
